packages/synapse-react-client/src/components/EntityUpload/EntityUpload.tsx

```diff
@@ -0,0 +1,321 @@
+                }
+              }}
+              // @ts-expect-error - webkitdirectory is not included in the React.InputHTMLAttributes type
+              webkitdirectory="true"
+            />
+
+            <Typography variant={'body1'} my={2}>
+              <Link onClick={handleClick}>Click to upload</Link> or drag and
+              drop
+            </Typography>
+            <Menu
+              anchorEl={anchorEl}
+              open={open}
+              onClose={handleClose}
+              anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
+              transformOrigin={{ vertical: 'top', horizontal: 'left' }}
+              TransitionComponent={Fade}
+            >
+              <MenuItem
+                onClick={() => {
+                  handleClose()
+                  fileInputRef.current!.click()
+                }}
+              >
+                Files
+              </MenuItem>
+              <MenuItem
+                onClick={() => {
+                  handleClose()
+                  folderInputRef.current!.click()
+                }}
+              >
+                Folder
+              </MenuItem>
+            </Menu>
+
+            <Typography variant={'smallText1'}>
+              All uploaded files will be stored in
+              {uploadDestination?.storageLocationId ===
+                SYNAPSE_STORAGE_LOCATION_ID && ' Synapse storage'}
+              {uploadDestination?.storageLocationId !==
+                SYNAPSE_STORAGE_LOCATION_ID && (
+                <>
+                  :<br />
+                  {uploadDestination?.banner}
+                </>
+              )}
+            </Typography>
+          </>
+        )}
+      </Stack>
+      {uploadProgress.length > 0 && (
+        <Paper
+          sx={{
+            px: `${UPLOAD_CONTAINER_PADDING_X_PX}px`,
+            py: 2,
+            mt: 3,
+            width: '100%',
+            height: '260px',
+            background: '#FBFBFC',
+          }}
+        >
+          <Typography variant={'headline2'} gutterBottom>
+            {state === 'WAITING' && <>Uploads</>}
+            {state === 'UPLOADING' && (
+              <>
+                Uploading {activeUploadCount}{' '}
+                {pluralize('Item', activeUploadCount)}
+              </>
+            )}
+            {state === 'COMPLETE' && (
+              <>
+                Uploaded {numberOfItemsCompleted}{' '}
+                {pluralize('Item', numberOfItemsCompleted)}
+              </>
+            )}
+          </Typography>
+          <Box
+            sx={{
+              '& > div > div > :not(:last-child)': {
+                borderBottom: 'solid 1px #EAECEE',
+              },
+            }}
+            // Add negative margin equivalent to the container's padding
+            // This will put the scrollbar on the right edge of the container
+            mx={`${UPLOAD_CONTAINER_PADDING_X_PX * -1}px`}
+          >
+            <FixedSizeList
```

Comment:
Virtualized the list of uploads with react-window since this may be really large, e.g. if you upload a folder containing hundreds of files